include/rdma/providers/fi_log.h

```diff
@@ -100,6 +100,16 @@ void fi_log(const struct fi_provider *prov, enum fi_log_level level,
 	do {} while (0)
 #endif
 
+#define FI_WARN_ONCE(prov, subsystem, ...) ({				\
+	static int warned;						\
+	int log_enabled = fi_log_enabled(prov, FI_LOG_WARN, subsystem);	\
```

Comment:
Will fi_log_enabled() ever change between successive calls?

Reply:
fi_log_enabled() won't change between successive calls. We need this on the first time call to see if the user enables warn level log though.
Are you concerning that if we can avoid this on the successive calls?

Reply:
I was wondering if we could change this to re-use the FI_WARN macro.
```
static int warned;
if (!warned) {
    FI_WARN(...)
    warned = 1;
}
```
Your method is technically more correct in that we only set warned after actually warning about it, but it does duplicate code and is slightly less efficient.

Reply:
I was following the way kernel code does, where I use `fi_log_enabled` as the condition. But yeah, it duplicates the code and is less efficient.

Another way I can think of is the following. We set warned only when warning and fi_log_enabled won't be evaluated if it's already warned. Duplicate code is still there though. 
```
static int warned; 
if (!warned && fi_log_enabled(prov, FI_LOG_WARN, subsystem)) {
        fi_log(prov, FI_LOG_WARN, subsystem, ...); 
        warned = 1; 
}  
```